rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/TypeSupport.hpp

```diff
@@ -113,6 +119,10 @@ class TypeSupport : public eprosima::fastdds::dds::TopicDataType
 
   bool max_size_bound_;
   bool is_plain_;
+  bool key_is_unbounded_;
+  mutable size_t key_max_serialized_size_;
+  mutable MD5 md5_;
+  mutable std::vector<uint8_t> key_buffer_;
```

Comment:
I am worried about the possibility of multiple threads accessing these fields. I'd add a mutex and protect the access.

Think about the case of two RMW publishers with the same keyed type doing rmw_publish from different threads.

Reply:
The `getKey()` method is not protected upwards, so it makes sense.
What do you think on protecting the whole `get_key_hash_from_ros_message()` ? those fields are intensively used inside that method and I dont think that locking in a more granular way within the method would make it more performant

Reply:
> What do you think on protecting the whole `get_key_hash_from_ros_message()` ?

This is in fact what I thought, yes. Adding a `lock_guard` [here](https://github.com/eProsima/rmw_fastrtps/blob/21235c77b605882400dfea2979bfb23166a3a273/rmw_fastrtps_shared_cpp/src/TypeSupport_impl.cpp#L87)

Reply:
Exactly, that what I was thinking too, adding it